hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala

```diff
@@ -0,0 +1,281 @@
+   *
+   * @param taskContext task context
+   * @param dir         optional subdirectory
+   * @param ext         file extension
+   * @return a path as a string
+   */
+  override def newTaskTempFile(
+      taskContext: TaskAttemptContext,
+      dir: Option[String],
+      ext: String): String = {
+
+    val workDir = committer.getWorkPath
+    val parent = dir.map(d => new Path(workDir, d)).getOrElse(workDir)
+    val file = new Path(parent, buildFilename(taskContext, ext))
+    logDebug(s"Creating task file $file for dir $dir and ext $ext")
+    file.toString
+  }
+
+  /**
+   * Absolute files are still renamed into place with a warning.
+   *
+   * The created file will be committed through the `rename()` operation
+   * of the filesystem/filestore connector, which may be a slow `O(data)`
+   * copy. Because directory listings are *not* used to determine which files
+   * to commit, even when a store has listing inconsistency (i.e. Amazon S3),
+   * there is no risk of that affecting the operation.
+   * Similarly, because the superclass's implementation of this method uses
+   * a UUID in the source name, eventual consistency on an update will not be
+   * an issue.
+   * There is a very small risk that Amazon S3's brief caching of 404s
+   * can cause any probes for the file to fail if the task is committed
+   * immediately after the file was created.
+   *
+   * @param taskContext task
+   * @param absoluteDir destination dir
+   * @param ext         extension
+   * @return an absolute path
+   */
+  override def newTaskTempFileAbsPath(
+    taskContext: TaskAttemptContext,
+    absoluteDir: String,
+    ext: String): String = {
+
+    val file = super.newTaskTempFileAbsPath(taskContext, absoluteDir, ext)
+    logInfo(s"Creating temporary file $file for absolute path for dir $absoluteDir")
+    file
+  }
+
+  /**
+   * Build a filename which is unique across all task events.
+   * It does not have to be consistent across multiple attempts of the same
+   * task or job.
+   *
+   * @param taskContext task context
+   * @param ext         extension
+   * @return a name for a file which must be unique across all task attempts
+   */
+  protected def buildFilename(
+    taskContext: TaskAttemptContext,
+    ext: String): String = {
+
+    // The file name looks like part-00000-2dd664f9-d2c4-4ffe-878f-c6c70c1fb0cb_00003-c000.parquet
+    // Note that %05d does not truncate the split number, so if we have more than 100000 tasks,
+    // the file name is fine and won't overflow.
+    val split = taskContext.getTaskAttemptID.getTaskID.getId
+    f"part-$split%05d-$jobId$ext"
+  }
+
+  override def setupJob(jobContext: JobContext): Unit = {
+    logDebug("setup job")
+    super.setupJob(jobContext)
+  }
+
+  override def commitJob(
+      jobContext: JobContext,
+      taskCommits: Seq[FileCommitProtocol.TaskCommitMessage]): Unit = {
+    logDebug(s"commit job with ${taskCommits.length} task commit message(s)")
```

Comment:
Are these overrides just to add a debug message really helpful?

If so, why not have the debug message in the parent class?

Reply:
It is handy to have the option of debugging what is going on, and I was trying to be as minimally invasive to the existing code as possible. But given their value, pulling them up is a good idea. Will do that and then delete these methods

Reply:
Did you mean you were going to change the parent class as part of this change or separately?

Reply:
pulling up. FWIW, If I were going to make bigger changes to the parent class it'd be

* to actually log the duration of renames
* consider actually doing them in parallel. Not just for performance against slow stores (s3, openstack), but to avoid renaming large files triggering timeouts. I've been playing a lot with rename perf recently (HADOOP-15183) and even with parallelized renaming of individual blocks there's scope for speedup
* Track which parent partition paths have already been created, so skip all the exists() checks. Saves 3 S3 requests per file.

That will be something for a new JIRA